Minor refactoring of assertion code

pvollan opened the pull request:
#### 09ce8066c4df92cbcb1ba906ffaea4bc76c4aab8
<pre>
Minor refactoring of assertion code
<a href="https://bugs.webkit.org/show_bug.cgi?id=263092">https://bugs.webkit.org/show_bug.cgi?id=263092</a>
rdar://116884036

Reviewed by Brent Fulgham.

This patch makes AuxiliaryProcessProxy inherit from ProcessThrottlerClient, so that every process proxy class
does not need to do that. Additionally, a new create method is added to ProcessAssertion, which takes a
AuxiliaryProcessProxy parameter, instead of a PID. This static method is intended to be used when you want to
take an assertion on the GPU, Networking, or WebContent process from the UI process. This patch is also
preparing for adopting new assertion SPI from ExtensionKit, but this is not enabled by default, so there
should be no behavior change from this patch.

* Source/WebKit/NetworkProcess/Downloads/DownloadMap.cpp:
(WebKit::DownloadMap::add):
* Source/WebKit/Platform/spi/Cocoa/ExtensionKitSPI.h:
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
* Source/WebKit/UIProcess/Cocoa/AuxiliaryProcessProxyCocoa.mm:
* Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm:
(WebKit::ProcessAssertion::ProcessAssertion):
(WebKit::ProcessAssertion::~ProcessAssertion):
* Source/WebKit/UIProcess/Downloads/DownloadProxyMap.cpp:
(WebKit::DownloadProxyMap::createDownloadProxy):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::setWebProcessHasUploads):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/ProcessAssertion.cpp:
(WebKit::ProcessAssertion::create):
* Source/WebKit/UIProcess/ProcessAssertion.h:
(WebKit::aquireAssertion):
(WebKit::ProcessAssertion::create): Deleted.
* Source/WebKit/UIProcess/ProcessThrottlerClient.h:
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/269791@main">https://commits.webkit.org/269791@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f8fc1230061a774b0f9ba99408f7bea5634b142e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23477 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1593 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24600 "Built successfully") | [⏳ 🛠 wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21673 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/3144 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/24018 "Built successfully") | [⏳ 🧪 wpe-wk2 ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23719 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/1139 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20315 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26234 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/941 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21213 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/20430 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/21396 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21474 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/25253 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/22828 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/923 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18646 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/30215 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/892 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/20988 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/6234 "Passed tests") | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5633 "Failed to canonicalize commit") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1334 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/30167 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3024 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1198 "Built successfully") | | [❌ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/6128 "Found 1 new JSC stress test failure: microbenchmarks/array-from-object-func.js.lockdown (failure)") | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm:
```diff
@@ -389,6 +393,22 @@ - (void)assertion:(RBSAssertion *)assertion didInvalidateWithError:(NSError *)er
+    NSString *runningBoardAssertionName = runningBoardNameForAssertionType(assertionType);
+    NSString *runningBoardDomain = runningBoardDomainForAssertionType(assertionType);
+    auto capability = adoptNS([_SECapabilities assertionWithDomain:runningBoardDomain name:runningBoardAssertionName]);
+    _SEExtensionProcess *extensionProcess = process.extensionProcess();
```
Doesn't `process.extensionProcess()` return a `RetainPtr<_SEExtensionProcess>`?

Seems like this should be `auto extensionProcess = process.extensionProcess()`

brentfulgham commented on Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm:
```diff
@@ -389,6 +393,22 @@ - (void)assertion:(RBSAssertion *)assertion didInvalidateWithError:(NSError *)er
+    NSString *runningBoardDomain = runningBoardDomainForAssertionType(assertionType);
+    auto capability = adoptNS([_SECapabilities assertionWithDomain:runningBoardDomain name:runningBoardAssertionName]);
+    _SEExtensionProcess *extensionProcess = process.extensionProcess();
+    m_grant = [extensionProcess grantCapabilities:capability.get() error:nil];
```
I think it would be useful to log if the `grantCapabilities` operation fails, since we often have bugs about background processes or other things being slowed down or terminated while doing useful things. If this operation fails, we should log it to help debug things in the future.

We should pass an NSError into `error:` and if `m_grant` is `nil` after this, we should log what the error says.

brentfulgham commented on Source/WebKit/UIProcess/ProcessAssertion.cpp:
```diff
@@ -52,6 +53,32 @@ ASCIILiteral processAssertionTypeDescription(ProcessAssertionType type)
+Ref<ProcessAssertion> ProcessAssertion::create(AuxiliaryProcessProxy& process, const String& reason, ProcessAssertionType type, Mode mode, CompletionHandler<void()>&& acquisisionHandler)
+{
+    auto assertion = adoptRef(*new ProcessAssertion(process, reason, type));
+    if (mode == Mode::Async)
```
All the code from here on is identical to the `ProcessID` version of the create method. I think you should move this common initialization code into a helper (maybe `createCommon` or something) that would share that logic.

pvollan replied:
Good catches! I have resolved these issues in the latest patch.

Thanks for reviewing!

brentfulgham commented on Source/WebKit/UIProcess/ProcessThrottler.cpp:
```diff
@@ -301,6 +306,17 @@ void ProcessThrottler::updateThrottleStateIfNeeded()
+#if USE(EXTENSIONKIT_ASSERTIONS)
+void ProcessThrottler::didConnectToProcess(RetainPtr<_SEExtensionProcess> process)
+{
+    PROCESSTHROTTLER_RELEASE_LOG("didConnectToProcess");
```
It's too bad `_SEExtensionProces` doesn't give us a way to see the PID of the underlying process. Logging it here would be helpful for debugging.

brentfulgham reviewed (approved):
I think you are assigning a RetainPtr to a bare pointer, which seems wrong. Otherwise, the patch looks good. r=me with that fix.

pvollan committed:
Minor refactoring of assertion code
https://bugs.webkit.org/show_bug.cgi?id=263092
rdar://116884036

Reviewed by Brent Fulgham.

This patch makes AuxiliaryProcessProxy inherit from ProcessThrottlerClient, so that every process proxy class
does not need to do that. Additionally, a new create method is added to ProcessAssertion, which takes a
AuxiliaryProcessProxy parameter, instead of a PID. This static method is intended to be used when you want to
take an assertion on the GPU, Networking, or WebContent process from the UI process. This patch is also
preparing for adopting new assertion SPI from ExtensionKit, but this is not enabled by default, so there
should be no behavior change from this patch.

* Source/WebKit/NetworkProcess/Downloads/DownloadMap.cpp:
(WebKit::DownloadMap::add):
* Source/WebKit/Platform/spi/Cocoa/ExtensionKitSPI.h:
* Source/WebKit/UIProcess/AuxiliaryProcessProxy.h:
* Source/WebKit/UIProcess/Cocoa/AuxiliaryProcessProxyCocoa.mm:
* Source/WebKit/UIProcess/Cocoa/ProcessAssertionCocoa.mm:
(WebKit::ProcessAssertion::ProcessAssertion):
(WebKit::ProcessAssertion::~ProcessAssertion):
* Source/WebKit/UIProcess/Downloads/DownloadProxyMap.cpp:
(WebKit::DownloadProxyMap::createDownloadProxy):
* Source/WebKit/UIProcess/GPU/GPUProcessProxy.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::setWebProcessHasUploads):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/ProcessAssertion.cpp:
(WebKit::ProcessAssertion::create):
* Source/WebKit/UIProcess/ProcessAssertion.h:
(WebKit::aquireAssertion):
(WebKit::ProcessAssertion::create): Deleted.
* Source/WebKit/UIProcess/ProcessThrottlerClient.h:
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/269791@main
